haystack/components/preprocessors/markdown_header_splitter.py

```diff
@@ -0,0 +1,405 @@
+
+                # preserve header metadata
+                for key in ["header", "parentheaders"]:
+                    if key in doc.meta:
+                        split.meta[key] = doc.meta[key]
+
+                # preserve primary split ID
+                if "split_id" in doc.meta:
+                    split.meta["header_split_id"] = doc.meta["split_id"]
+
+                result_docs.append(split)
+
+        logger.info(f"Secondary splitting complete. Final count: {len(result_docs)} documents.")
+        return result_docs
+
+    @component.output_types(documents=List[Document])
+    def run(self, documents: List[Document], infer_header_levels: Optional[bool] = None) -> Dict[str, List[Document]]:
+        """
+        Run the markdown header splitter with optional secondary splitting.
+
+        :param documents: List of documents to split
+        :param infer_header_levels: If True, attempts to infer and rewrite header levels before splitting.
+            If None, uses the value from initialization.
+        """
+        infer_header_levels = infer_header_levels if infer_header_levels is not None else self.infer_header_levels
+
+        # process documents - preprocess if told to
+        processed_documents = []
+        for doc in documents:
+            if infer_header_levels:
+                content = self._infer_and_rewrite_header_levels(doc.content)
+                processed_documents.append(Document(content=content, meta=doc.meta, id=doc.id))
+            else:
+                processed_documents.append(doc)
+
+        # split by markdown headers
+        header_splitter = CustomDocumentSplitter(
+            split_by="function",
+            splitting_function=lambda text: self._split_by_markdown_headers(text),
+            page_break_character=self.page_break_character,
+        )
+
+        # get splits
+        header_split_docs = header_splitter.run(documents=processed_documents)["documents"]
+        logger.info(f"Header splitting produced {len(header_split_docs)} documents")
+
+        # apply secondary splitting if requested
+        if self.secondary_split != "none":
+            final_docs = self._apply_secondary_splitting(header_split_docs)
+        else:
+            final_docs = header_split_docs
+
+        return {"documents": final_docs}
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Serialize component to dictionary."""
+        return default_to_dict(
+            self,
+            infer_header_levels=self.infer_header_levels,
+            page_break_character=self.page_break_character,
+            secondary_split=self.secondary_split,
+            split_length=self.split_length,
+            split_overlap=self.split_overlap,
+            split_threshold=self.split_threshold,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "MarkdownHeaderSplitter":
+        """Deserialize component from dictionary."""
+        return default_from_dict(cls, data)
```

Comment:
These methods aren't actually necessary to get sede to work for your components since all values are auto supported by our `component_to_dict` and `component_from_dict` methods which are used by `Pipeline` to serialize and deserialize components. 

Typically you only make a custom one when there is something non-standard to serialize like an api key or a function.